### ISSUE TITLE
Evil Dr. BMG MK3 Extruder Testing

### ISSUE THREAD

**Issue by megabytemike** [TheZeroBeast/MM-control-01#31]
I'm having an issue after moving from the standard Prusa firmware to the latest version of TheZeroBeast. When loading a filament to nozzle or when it loads itself via a gcode, it selects the correct filament then loads it perfectly all the way down to the extruder but then it stops and starts flashing red on the currently selected filament and errors out, it clears when i manually remove the filament from the extruder back past the selector (finda) then it loads it again and same error. I can not get it to get past this error, not sure if I did something wrong when flashing this firmware. I have the extruder Fil sensor on and have changed my Slic3r settings to match the readme. (MK3)...Is it supposed to load the filament past the sensor or stop before it? Mine is loading to the extruder gears like in the original firmware.

update, checked extruder gears during load and they never turn to grab the filament once the filament reaches the bondtech. I've also noticed after the firmware change, the Fil Sensor is defaulted to off and has to be turned on after every hard reset. I have also made sure the fil sensor is clean, even removed it to inspect it and clean it with iso.

**Comment by megabytemike** [comment]
update, checked extruder gears during load and they never turn to grab the filament once the filament reaches the bondtech.

**Comment by TheZeroBeast** [comment]
Hi Megabytemike,

Thanks for posting your issue. I haven't seen this one before.

Just to confirm, you have flashed my corresponding MK3 firmware as well as the MMU2 firmware?

**Comment by Evildr14** [comment]
I to am having the same issue. The filament sensor in the MK3 is not picking up the filament. 

23:19:10.323: tmc2130_goto_step 0 29 2 1000
23:19:10.412: tmc2130_home_exit tmc2130_sg_homing_axes_mask=0x01
23:19:10.532: tmc2130_home_enter(axes_mask=0x02)
23:19:15.270: 0 step=33 mscnt= 541
23:19:15.274: tmc2130_goto_step 1 33 2 1000
23:19:15.283: tmc2130_home_exit tmc2130_sg_homing_axes_mask=0x02
23:19:45.037: Unknown G code: $296 G21
23:19:45.039: MMU <= 'F0 0'
23:19:45.072: MMU <= 'F1 0'
23:19:45.207: MMU <= 'F2 0'
23:19:45.340: MMU <= 'F3 0'
23:19:45.443: MMU <= 'F4 2'
23:19:45.720: MMU => 'CMD ACK 0x11'
23:19:45.724: MMU <= 'T1'
23:19:46.281: MMU => 'waiting for filament @ MK3 Sensor'
23:21:25.553: MMU => 'CMD RETRY 0x81' (85)
23:21:25.973: INT4

I have will swap the sensor out of my second MK# once it finishes printing to see if that makes any difference.
![img_20181201_092118](https://user-images.githubusercontent.com/38388948/49320898-03847000-f550-11e8-8163-d323f63fdf26.jpg)

update, I have tried 5 different colours and found the white loads perfectly. I have cleaned the sensor and makes no difference. (grey, blue, red, orange and white PLA)

23:34:52.654: MMU => 'CMD ACK 0x10'
23:34:52.694: MMU <= 'T0'
23:34:52.749: MMU => 'waiting for filament @ MK3 Sensor'
23:35:02.088: MMU => 'CMD ACK 0x81'
23:35:02.091: MMU <= 'Filament seen at extruder'
23:35:03.090: MMU => 'ok :)'
23:35:03.092: MMU => 'ok'

**Comment by AbeFM** [comment]
That was my experience. I attributed it to broken sensor, but it worked in
other cases.

**Comment by Evildr14** [comment]
I changed out the filament sensor and i can get orange and white to work but all others still fail. Not to sure what i can do now....
![img_20181201_102705](https://user-images.githubusercontent.com/38388948/49321824-d9ce4780-f555-11e8-895e-2029204c1535.jpg)

**Comment by megabytemike** [comment]
Yes, I flashed both several times and went back to stock to verify my sensor was working. I even tried the earlier release with the same problem. Wonder if it has something to do with the different sensor revisions.

**Comment by Evildr14** [comment]
I just started printing the indirect sensor mod from martin_au on thingiverse (https://www.thingiverse.com/thing:3223513). I'm not sure if it will change the filament sensing distance but I'm keen to give it a go.

**Comment by AbeFM** [comment]
I'm going to try th same - if nothing else, it'll keep me from breaking more sensors.

I wonder what the difference between the versions two is.

I'll use the external one as it's already set up for a passthrough, and they get around fliping by mounting sensor upside down. It'll change the distance from sensor to gears, though.

**Comment by TheZeroBeast** [comment]
That bearing sensor mod is a great idea, keen to do it when I find time. I hear PRUSA has a new sensor they're working on, hopefully we MK3 owners will get it complementary or even first go at them if it gets a solid filament sensor. PAT9130EW I do believe.

**Comment by Evildr14** [comment]
Well I'm almost set. I have printed out the parts for the bearing mod just waiting on the bearings. With the inconsistencies of the current setup I dare say they will be putting a bit of effort into it. Unfortunately  with the number of MK3s in the world now, anything they release would take months to arrive.

**Comment by AbeFM** [comment]
I found the bearing to be loose in the housing, so I'm trying to narrow it up, and redo the top to get a better bowden tube set up.

**Comment by Evildr14** [comment]
@AbeFM Nice. let me know how you go.

**Comment by AbeFM** [comment]
I'm using this one: https://www.thingiverse.com/thing:3129921 - it will definately need the distance from sensor to pullys changesd, so either way there is a firmware change, I guess.
It keeps me from having to reprint the housing or tweak the firmware. I have to admit, once I get into Git more than "commit" or "pull" it stops makes sense and starts taking hours. I've been trying to get the commits you mentioned but it has become a lost cause.

**Comment by Evildr14** [comment]
My understanding of git is rather limited at this stage. I tried to build the indirect filament sensor of thingiverse without much success. If the bearing is too close the bearing will turn but won't let the tip through and if it is too far away the bearing will slip on the filament. I started designing my own which seems to work quite well but adds a little weight. I figure if I'm going to change the firmware I might as well add a BMG extruder.

**Comment by Evildr14** [comment]
![img_20181206_090525](https://user-images.githubusercontent.com/38388948/49639524-f509ed00-fa56-11e8-9bd1-b5b075bccb5e.jpg)
![img_20181206_090437](https://user-images.githubusercontent.com/38388948/49639539-00f5af00-fa57-11e8-9be4-f27f647ca04f.jpg)

**Comment by TheZeroBeast** [comment]
How goes your new sensor setup Evildr14?

**Comment by Evildr14** [comment]
Slow and steady progress, my final bits arrived in the mail today...the new pancake motor. I almost had tears drilling out the new BMG to allow filament tips through....the tolerance of the bmg is way too tight.

**Comment by Evildr14** [comment]
![img_20181211_194741](https://user-images.githubusercontent.com/38388948/49792376-4cbd9680-fd7e-11e8-9d00-9401258b8d36.jpg)
![img_20181211_194720](https://user-images.githubusercontent.com/38388948/49792392-58a95880-fd7e-11e8-819e-26b44b384743.jpg)
![Uploading IMG_20181211_194716.jpg…]()

**Comment by TheZeroBeast** [comment]
Looking good none the less but yes, it would of been hard drilling something so nice and new.

Also, this issue thread is now resolved with the latest RC code. I will rename so you can share your progress with the BMG.

**Comment by AbeFM** [comment]
To set Optical Filament Sensor to Bondetch Distance, I modified line 32 of Motion.cpp:
const int STEPS_MK3FSensor_To_Bondtech = 756; //Abe 385 original number, I need to modify 385 by 191% (32mm/16.5mm) = 755.xxx

Just to tie up the thought.

Looks like an interesting build. I found I had to the the filament/bearing sensor just right as well. There are grooved pulleys, and I think this is the way to go - less deflection!

**Comment by Evildr14** [comment]
Thanks for that, that will definitely save me a bit Time in the firmware. 

The grove in the pulleys grabbed the filament a lot better than the bearing, which meant I didn't need much deflection to turn it. It adds weight but it seems to be able to handle the filament tips as well. I'm keen to test run this weekend.

**Comment by TheZeroBeast** [comment]
Thanks for the update, keen for the test print results 👍

**Comment by AbeFM** [comment]
Did you try a grooved pulley? I'm pretty convinced but haven't built one yet.

**Comment by Evildr14** [comment]
Sorry for the slow updates, half way through my rebuild bi broke my thermistor wires.... 😢...now I am waiting for a new one to arrive to do the test run....and being Xmas, I don't expect it to arrive till the end of the month...I will upload the step files to thingiverse tonight if anyone would like to have a look....

**Comment by AbeFM** [comment]
Get something generic on Amazon? The last time I ordered one, I ordered 2 just in case. And put a connector on for easier changes. :-)

**Comment by Evildr14** [comment]
I will definitely order a couple this time round. https://www.thingiverse.com/thing:3299790 is the files if anyone is keen for a look.

**Comment by Mjankor** [comment]
I had this bug. To resolve it I disconnected the MMU and checked that the printer was working properly under normal operation (so checked autoload and filament sensor status). I also disabled and re-enabled the filament sensor. I then plugged the MMU back in and reflashed it. This then worked, with "triggered" showing up on feeding filament into the top of the extruder. I think that when the MMU is connected, the filament sensor is still disabled, even if it says it's on in the menu.
